File: packages/ts/src/prompt.ts
import { EMBEDDED_CAPABILITY_REGISTRY, EMBEDDED_PROMPT_FRAGMENTS } from "./prompt-data.js";
import { EXTRACTION_CAPABILITIES, type ExtractionCapability } from "./schema.js";

export type PromptProfile = "minimal" | "standard" | "full";

export interface CapabilityDefinition {
  name: ExtractionCapability;
  depends_on?: ExtractionCapability[];
  embedding_input?: string;
  base?: boolean;
  modifier_only?: boolean;
  rule?: string;
}

export interface CapabilityRegistry {
  version: "1";
  profiles: Record<PromptProfile, ExtractionCapability[]>;
  omit_when_absent: ExtractionCapability[];
  standard_embedding_inputs: string[];
  capabilities: CapabilityDefinition[];
}

function duplicateValues(values: string[]): string[] {
  return values.filter((value, index) => values.indexOf(value) !== index);
}

function validateCapabilityRegistry(registry: CapabilityRegistry, sourcePath: string): void {
  const schemaCapabilities = new Set<string>(EXTRACTION_CAPABILITIES);
  const names = registry.capabilities.map((definition) => definition.name);
  const registryCapabilities = new Set<string>(names);
  const duplicateCapabilities = duplicateValues(names);
  if (duplicateCapabilities.length > 0) {
    throw new Error(`Capability registry has duplicate capabilities: ${[...new Set(duplicateCapabilities)].sort().join(", ")}`);
  }

  const unknown = names.filter((name) => !schemaCapabilities.has(name));
  if (unknown.length > 0) {
    throw new Error(`Capability registry contains unknown capabilities: ${unknown.sort().join(", ")}`);
  }

  const missing = EXTRACTION_CAPABILITIES.filter((name) => !registryCapabilities.has(name));
  if (missing.length > 0) {
    throw new Error(`Capability registry is missing schema capabilities: ${missing.sort().join(", ")}`);
  }

  for (const definition of registry.capabilities) {
    for (const dep of definition.depends_on ?? []) {
      if (!registryCapabilities.has(dep)) {
        throw new Error(`Capability registry dependency ${definition.name}.${dep} is not a valid capability`);
      }
    }
  }

  for (const [profile, capabilities] of Object.entries(registry.profiles) as [PromptProfile, ExtractionCapability[]][]) {
    for (const capability of capabilities) {
      if (!registryCapabilities.has(capability)) {
        throw new Error(`Profile ${profile} references unknown capability ${capability}`);
      }
    }
  }

  for (const capability of registry.omit_when_absent) {
    if (!registryCapabilities.has(capability)) {
      throw new Error(`omit_when_absent references unknown capability ${capability}`);
    }
  }

  if (!Array.isArray(registry.standard_embedding_inputs)) {
    throw new Error("Capability registry standard_embedding_inputs must be a list");
  }
  const embeddingInputs = new Set<string>([
    "source",
    ...registry.capabilities
      .map((definition) => definition.embedding_input)
      .filter((input): input is string => typeof input === "string"),
  ]);
  for (const input of registry.standard_embedding_inputs) {
    if (!embeddingInputs.has(input)) {
      throw new Error(`standard_embedding_inputs references unknown input ${input}`);
    }
  }

  if (registry.version !== "1") {
    throw new Error(`Unsupported capability registry version in ${sourcePath}: ${registry.version}`);
  }
}

function loadCapabilityRegistry(): CapabilityRegistry {
  const registry = EMBEDDED_CAPABILITY_REGISTRY;
  validateCapabilityRegistry(registry, "embedded capability registry");
  return registry;
}

export const CAPABILITY_REGISTRY = loadCapabilityRegistry();
export const CANONICAL_ORDER: ExtractionCapability[] = CAPABILITY_REGISTRY.capabilities.map((definition) => definition.name);
export const VALID_CAPABILITIES: Set<string> = new Set(CANONICAL_ORDER);
export const BASE_CAPABILITIES: Set<string> = new Set(
  CAPABILITY_REGISTRY.capabilities
    .filter((definition) => definition.base === true)
    .map((definition) => definition.name),
);
export const MODIFIER_ONLY: Set<string> = new Set(
  CAPABILITY_REGISTRY.capabilities
    .filter((definition) => definition.modifier_only === true)
    .map((definition) => definition.name),
);
export const CAPABILITY_DEPS: Partial<Record<ExtractionCapability, ExtractionCapability[]>> = Object.fromEntries(
  CAPABILITY_REGISTRY.capabilities
    .filter((definition) => definition.depends_on !== undefined)
    .map((definition) => [definition.name, definition.depends_on ?? []]),
) as Partial<Record<ExtractionCapability, ExtractionCapability[]>>;
export const CAPABILITY_RULES: Partial<Record<ExtractionCapability, string>> = Object.fromEntries(
  CAPABILITY_REGISTRY.capabilities
    .filter((definition) => definition.rule !== undefined)
    .map((definition) => [definition.name, definition.rule ?? ""]),
) as Partial<Record<ExtractionCapability, string>>;
export const OMIT_WHEN_ABSENT: ExtractionCapability[] = CAPABILITY_REGISTRY.omit_when_absent;
export const STANDARD_EMBEDDING_INPUTS: string[] = CAPABILITY_REGISTRY.standard_embedding_inputs;
const CAPABILITY_EMBEDDING_INPUTS: Partial<Record<ExtractionCapability, string>> = Object.fromEntries(
  CAPABILITY_REGISTRY.capabilities
    .filter((definition) => definition.embedding_input !== undefined)
    .map((definition) => [definition.name, definition.embedding_input ?? ""]),
) as Partial<Record<ExtractionCapability, string>>;

export interface PromptOptions {
  capabilities?: CapabilityInput[];
  profile?: PromptProfile;
  add?: CapabilityInput[];
  remove?: CapabilityInput[];
  categories?: string[];
  source_type?: string;
  date?: string;
  stage?: "stage1";
  extracted_at?: string;
  compact?: boolean;
  embed?: Partial<Record<ExtractionCapability, boolean>>;
}

export interface CapabilitySpec {
  name?: ExtractionCapability;
  capability?: ExtractionCapability;
  embed?: boolean;
  embedding?: boolean;
}

export type CapabilityInput = ExtractionCapability | CapabilitySpec;

function buildRunConstraintRules(resolved: ExtractionCapability[], options: PromptOptions): string[] {
  const rules: string[] = [];
  const capabilities = new Set(resolved);

  if (options.compact) {
    rules.push("Keep this extraction compact and high signal.");
  }
  if (options.extracted_at) {
    rules.push(`Use exactly this extracted_at value: ${options.extracted_at}.`);
  }
  if (options.stage === "stage1") {
    rules.push(
      "Produce Stage 1 content only. Do not include version, produced_by, source_id, source_type, kind, capabilities, extensions, or embeddings.",
    );
    rules.push("Only include fields represented in the requested capability set and response schema.");
  }
  if (capabilities.has("entity_ids")) {
    rules.push('Entity IDs are extraction-local only. Use short IDs like "e1", "e2", "e3".');
  }
  if (capabilities.has("goal_entity_refs")) {
    rules.push("Every goal.entity_refs entry must refer to one of the entity IDs you emit.");
  }
  if (!capabilities.has("temporal_refs")) {
    rules.push("Omit temporal_refs for this run.");
  }
  if (!capabilities.has("relations")) {
    rules.push("Omit relations for this run.");
  }
  if (!capabilities.has("assertion_signals")) {
    rules.push("Omit signals fields for this run.");
  }
  for (const capability of OMIT_WHEN_ABSENT) {
    if (!capabilities.has(capability)) {
      rules.push(`Omit ${capability} for this run.`);
    }
  }

  return rules;
}

function loadProfile(name: string): ExtractionCapability[] {
  if (!(name in CAPABILITY_REGISTRY.profiles)) {
    throw new Error(`Unknown profile: ${name}`);
  }
  return CAPABILITY_REGISTRY.profiles[name as PromptProfile];
}

export function profileCapabilities(profile: NonNullable<PromptOptions["profile"]>): ExtractionCapability[] {
  return [...loadProfile(profile)];
}

export function capabilityEmbeddingInput(capability: ExtractionCapability): string | undefined {
  return CAPABILITY_EMBEDDING_INPUTS[capability];
}

function loadFragment(name: string): string {
  const fragment = EMBEDDED_PROMPT_FRAGMENTS[name];
  if (fragment === undefined) {
    throw new Error(`Missing prompt fragment: ${name}`);
  }
  return fragment;
}

function renderTemplate(template: string, ctx: Record<string, unknown>): string {
  let result = template.replace(
    /\{\{#if (\w+)\}\}([\s\S]*?)\{\{\/if\}\}/g,
    (_match, varName: string, body: string) => {
      return ctx[varName] ? body : "";
    },
  );
  return renderVars(result, ctx);
}

function renderVars(template: string, ctx: Record<string, unknown>): string {
  return template.replace(/\{\{(\w+)\}\}/g, (_match, varName: string) => {
    const val = ctx[varName];
    if (Array.isArray(val)) return val.join(", ");
    return val != null ? String(val) : "";
  });
}

export function capabilityName(capability: CapabilityInput): ExtractionCapability {
  if (typeof capability === "string") return capability;
  const name = capability.name ?? capability.capability;
  if (typeof name !== "string") {
    throw new Error("Capability object must include name or capability");
  }
  return name;
}

export function normalizeCapabilityInputs(capabilities: Iterable<CapabilityInput>): ExtractionCapability[] {
  return [...capabilities].map(capabilityName);
}

export function capabilityRequestsEmbedding(capability: CapabilityInput): boolean {
  return typeof capability === "object" && capability !== null && (capability.embed === true || capability.embedding === true);
}

export function capabilityEmbeddingPreference(capability: CapabilityInput): boolean | undefined {
  if (typeof capability !== "object" || capability === null) return undefined;
  if (typeof capability.embed === "boolean") return capability.embed;
  if (typeof capability.embedding === "boolean") return capability.embedding;
  return undefined;
}

function validateCapabilityNames(caps: Iterable<CapabilityInput>, source: string): void {
  const unknown: string[] = [];
  for (const c of caps) {
    const name = capabilityName(c);
    if (!VALID_CAPABILITIES.has(name)) unknown.push(name);
  }
  if (unknown.length > 0) {
    throw new Error(`Unknown ${source}: ${unknown.sort().join(", ")}`);
  }
}

export function resolveCapabilities(options: Pick<PromptOptions, "capabilities" | "profile" | "add" | "remove">): ExtractionCapability[] {
  let caps: Set<ExtractionCapability>;

  if (options.capabilities != null) {
    validateCapabilityNames(options.capabilities, "capabilities");
    caps = new Set(normalizeCapabilityInputs(options.capabilities));
  } else if (options.profile != null) {
    caps = new Set(loadProfile(options.profile));
  } else {
    throw new Error("Either capabilities or profile must be provided");
  }

  if (options.add) {
    validateCapabilityNames(options.add, "capabilities in add");
    for (const c of normalizeCapabilityInputs(options.add)) caps.add(c);
  }
  const removed = new Set(options.remove ? expandCapabilityExclusions(normalizeCapabilityInputs(options.remove)) : []);

  let changed = true;
  while (changed) {
    changed = false;
    for (const cap of [...caps]) {
      const deps = CAPABILITY_DEPS[cap];
      if (deps) {
        for (const dep of deps) {
          if (!caps.has(dep)) {
            caps.add(dep);
            changed = true;
          }
        }
      }
    }
  }

  for (const c of removed) caps.delete(c);

  if (caps.size === 0) {
    throw new Error("Resolved capability set is empty");
  }

  const modifiers = [...caps].filter((c) => MODIFIER_ONLY.has(c));
  if (modifiers.length > 0 && ![...caps].some((c) => BASE_CAPABILITIES.has(c))) {
    throw new Error(
      `Modifier capabilities [${modifiers.sort().join(", ")}] require at least one ` +
      `base capability (${[...BASE_CAPABILITIES].sort().join(", ")})`
    );
  }

  return [...caps].sort((a, b) => {
    const ai = CANONICAL_ORDER.indexOf(a);
    const bi = CANONICAL_ORDER.indexOf(b);
    return (ai === -1 ? CANONICAL_ORDER.length : ai) - (bi === -1 ? CANONICAL_ORDER.length : bi);
  });
}

export function expandCapabilityExclusions(capabilities: Iterable<ExtractionCapability>): ExtractionCapability[] {
  const excluded = new Set(capabilities);
  let changed = true;
  while (changed) {
    changed = false;
    for (const capability of VALID_CAPABILITIES as Set<ExtractionCapability>) {
      if (excluded.has(capability)) continue;
      const deps = CAPABILITY_DEPS[capability] ?? [];
      if (deps.some((dep) => excluded.has(dep))) {
        excluded.add(capability);
        changed = true;
      }
    }
  }
  return [...excluded].sort((a, b) => CANONICAL_ORDER.indexOf(a) - CANONICAL_ORDER.indexOf(b));
}

export function buildExtractionPrompt(text: string, options: PromptOptions): string {
  if (options.capabilities != null && options.profile != null) {
    throw new Error("Cannot specify both capabilities and profile");
  }

  const resolved = resolveCapabilities(options);

  const ctx: Record<string, unknown> = {
    text,
    categories: options.categories,
    source_type: options.source_type,
    date: options.date,
  };

  const parts: string[] = [];

  const preamble = renderTemplate(loadFragment("preamble"), ctx);
  parts.push(preamble.trim());

  for (const cap of resolved) {
    const fragment = renderTemplate(loadFragment(cap), ctx);
    parts.push(fragment.trimEnd());
  }

  const rulesSection: string[] = [];
  for (const cap of resolved) {
    const rule = CAPABILITY_RULES[cap];
    if (rule) rulesSection.push(rule);
  }
  const runConstraintRules = buildRunConstraintRules(resolved, options);

  let postamble = renderTemplate(loadFragment("postamble"), ctx).trimEnd();
  if (rulesSection.length > 0 || runConstraintRules.length > 0) {
    const extraBlocks: string[] = [];
    if (rulesSection.length > 0) {
      extraBlocks.push(rulesSection.map((r) => `- ${r}`).join("\n"));
    }
    if (runConstraintRules.length > 0) {
      extraBlocks.push(`Additional run constraints:\n${runConstraintRules.map((r) => `- ${r}`).join("\n")}`);
    }
    const extraRules = extraBlocks.join("\n");
    const idx = postamble.indexOf("\nText:");
    if (idx >= 0) {
      postamble = postamble.slice(0, idx) + "\n" + extraRules + postamble.slice(idx);
    }
  }
  parts.push(postamble);

  return parts.join("\n") + "\n";
}

File: packages/python/src/synapt_extract/prompt.py
"""Composable prompt system for SynaptExtraction IL v1."""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import Any

from synapt_extract.schema import EXTRACTION_CAPABILITIES

_INSTALLED_PROMPTS = Path(__file__).resolve().parent / "prompts"
_REPO_PROMPTS = Path(__file__).resolve().parents[4] / "prompts"
PROMPTS_DIR = _INSTALLED_PROMPTS if _INSTALLED_PROMPTS.is_dir() else _REPO_PROMPTS


def _duplicates(values: list[str]) -> list[str]:
    return [value for index, value in enumerate(values) if value in values[:index]]


def _validate_capability_registry(registry: dict[str, Any], path: Path) -> None:
    if registry.get("version") != "1":
        raise ValueError(f"Unsupported capability registry version in {path}: {registry.get('version')}")

    definitions = registry.get("capabilities")
    if not isinstance(definitions, list):
        raise ValueError(f"Capability registry {path} must define capabilities")

    names = [definition.get("name") for definition in definitions if isinstance(definition, dict)]
    if len(names) != len(definitions) or not all(isinstance(name, str) for name in names):
        raise ValueError(f"Capability registry {path} contains malformed capability definitions")

    duplicates = _duplicates(names)
    if duplicates:
        raise ValueError(f"Capability registry has duplicate capabilities: {', '.join(sorted(set(duplicates)))}")

    registry_capabilities = set(names)
    unknown = registry_capabilities - EXTRACTION_CAPABILITIES
    if unknown:
        raise ValueError(f"Capability registry contains unknown capabilities: {', '.join(sorted(unknown))}")

    missing = EXTRACTION_CAPABILITIES - registry_capabilities
    if missing:
        raise ValueError(f"Capability registry is missing schema capabilities: {', '.join(sorted(missing))}")

    for definition in definitions:
        deps = definition.get("depends_on", [])
        if not isinstance(deps, list):
            raise ValueError(f"Capability registry dependency list for {definition.get('name')} must be a list")
        for dep in deps:
            if dep not in registry_capabilities:
                raise ValueError(f"Capability registry dependency {definition.get('name')}.{dep} is not valid")

    profiles = registry.get("profiles")
    if not isinstance(profiles, dict):
        raise ValueError(f"Capability registry {path} must define profiles")
    for profile, capabilities in profiles.items():
        if not isinstance(capabilities, list):
            raise ValueError(f"Profile {profile} capabilities must be a list")
        for capability in capabilities:
            if capability not in registry_capabilities:
                raise ValueError(f"Profile {profile} references unknown capability {capability}")

    omit_when_absent = registry.get("omit_when_absent", [])
    if not isinstance(omit_when_absent, list):
        raise ValueError("Capability registry omit_when_absent must be a list")
    for capability in omit_when_absent:
        if capability not in registry_capabilities:
            raise ValueError(f"omit_when_absent references unknown capability {capability}")

    standard_embedding_inputs = registry.get("standard_embedding_inputs")
    if not isinstance(standard_embedding_inputs, list):
        raise ValueError("Capability registry standard_embedding_inputs must be a list")
    embedding_inputs = {
        "source",
        *(
            definition.get("embedding_input")
            for definition in definitions
            if isinstance(definition.get("embedding_input"), str)
        ),
    }
    for input_name in standard_embedding_inputs:
        if input_name not in embedding_inputs:
            raise ValueError(f"standard_embedding_inputs references unknown input {input_name}")


def _load_capability_registry() -> dict[str, Any]:
    path = PROMPTS_DIR / "capabilities.json"
    if not path.exists():
        raise ValueError(f"Missing capability registry: {path}")
    registry = json.loads(path.read_text())
    _validate_capability_registry(registry, path)
    return registry


CAPABILITY_REGISTRY = _load_capability_registry()
_CAPABILITY_DEFINITIONS: list[dict[str, Any]] = CAPABILITY_REGISTRY["capabilities"]
CANONICAL_ORDER = [definition["name"] for definition in _CAPABILITY_DEFINITIONS]
VALID_CAPABILITIES = frozenset(CANONICAL_ORDER)
CAPABILITY_DEPS: dict[str, list[str]] = {
    definition["name"]: list(definition["depends_on"])
    for definition in _CAPABILITY_DEFINITIONS
    if definition.get("depends_on") is not None
}
CAPABILITY_RULES: dict[str, str] = {
    definition["name"]: definition["rule"]
    for definition in _CAPABILITY_DEFINITIONS
    if isinstance(definition.get("rule"), str)
}
BASE_CAPABILITIES = frozenset(
    definition["name"]
    for definition in _CAPABILITY_DEFINITIONS
    if definition.get("base") is True
)
MODIFIER_ONLY_CAPABILITIES = frozenset(
    definition["name"]
    for definition in _CAPABILITY_DEFINITIONS
    if definition.get("modifier_only") is True
)
CAPABILITY_EMBEDDING_INPUTS: dict[str, str] = {
    definition["name"]: definition["embedding_input"]
    for definition in _CAPABILITY_DEFINITIONS
    if isinstance(definition.get("embedding_input"), str)
}
OMIT_WHEN_ABSENT = tuple(CAPABILITY_REGISTRY["omit_when_absent"])
STANDARD_EMBEDDING_INPUTS = tuple(CAPABILITY_REGISTRY["standard_embedding_inputs"])


def _build_run_constraint_rules(
    resolved: list[str],
    *,
    stage: str | None = None,
    extracted_at: str | None = None,
    compact: bool | None = None,
) -> list[str]:
    rules: list[str] = []
    capabilities = set(resolved)

    if compact:
        rules.append("Keep this extraction compact and high signal.")
    if extracted_at:
        rules.append(f"Use exactly this extracted_at value: {extracted_at}.")
    if stage == "stage1":
        rules.append(
            "Produce Stage 1 content only. Do not include version, produced_by, source_id, "
            "source_type, kind, capabilities, extensions, or embeddings."
        )
        rules.append("Only include fields represented in the requested capability set and response schema.")
    if "entity_ids" in capabilities:
        rules.append('Entity IDs are extraction-local only. Use short IDs like "e1", "e2", "e3".')
    if "goal_entity_refs" in capabilities:
        rules.append("Every goal.entity_refs entry must refer to one of the entity IDs you emit.")
    if "temporal_refs" not in capabilities:
        rules.append("Omit temporal_refs for this run.")
    if "relations" not in capabilities:
        rules.append("Omit relations for this run.")
    if "assertion_signals" not in capabilities:
        rules.append("Omit signals fields for this run.")
    for capability in OMIT_WHEN_ABSENT:
        if capability not in capabilities:
            rules.append(f"Omit {capability} for this run.")

    return rules


def _load_profile(name: str) -> list[str]:
    profiles = CAPABILITY_REGISTRY["profiles"]
    if name not in profiles:
        raise ValueError(f"Unknown profile: {name}")
    return list(profiles[name])


def profile_capabilities(profile: str) -> list[str]:
    return list(_load_profile(profile))


def _load_fragment(name: str) -> str:
    path = PROMPTS_DIR / "v1" / f"{name}.txt"
    return path.read_text()


def _render_template(template: str, context: dict[str, Any]) -> str:
    def replace_if(match: re.Match) -> str:
        var = match.group(1)
        body = match.group(2)
        if context.get(var):
            return body
        return ""

    result = re.sub(r"\{\{#if (\w+)\}\}(.*?)\{\{/if\}\}", replace_if, template, flags=re.DOTALL)
    return _render_vars(result, context)


def _render_vars(template: str, context: dict[str, Any]) -> str:
    def replace_var(match: re.Match) -> str:
        var = match.group(1)
        val = context.get(var, "")
        if isinstance(val, list):
            return ", ".join(str(v) for v in val)
        return str(val)

    return re.sub(r"\{\{(\w+)\}\}", replace_var, template)


def capability_name(capability: Any) -> str:
    if isinstance(capability, str):
        return capability
    if isinstance(capability, dict):
        name = capability.get("name") or capability.get("capability")
        if isinstance(name, str):
            return name
    raise ValueError("Capability object must include name or capability")


def normalize_capability_inputs(capabilities: list[Any]) -> list[str]:
    return [capability_name(capability) for capability in capabilities]


def capability_requests_embedding(capability: Any) -> bool:
    return isinstance(capability, dict) and (capability.get("embed") is True or capability.get("embedding") is True)


def capability_embedding_preference(capability: Any) -> bool | None:
    if not isinstance(capability, dict):
        return None
    if isinstance(capability.get("embed"), bool):
        return capability["embed"]
    if isinstance(capability.get("embedding"), bool):
        return capability["embedding"]
    return None


def capability_embedding_input(capability: str) -> str | None:
    return CAPABILITY_EMBEDDING_INPUTS.get(capability)


def _validate_capability_names(caps: list[Any], source: str) -> None:
    names = set(normalize_capability_inputs(caps))
    unknown = names - VALID_CAPABILITIES
    if unknown:
        raise ValueError(f"Unknown {source}: {', '.join(sorted(unknown))}")


def resolve_capabilities(
    *,
    capabilities: list[Any] | None = None,
    profile: str | None = None,
    add: list[Any] | None = None,
    remove: list[Any] | None = None,
) -> list[str]:
    if capabilities is None and profile is None:
        raise ValueError("Either capabilities or profile must be provided")

    if capabilities is not None:
        _validate_capability_names(capabilities, "capabilities")
        caps = set(normalize_capability_inputs(capabilities))
    else:
        caps = set(_load_profile(profile))

    if add:
        _validate_capability_names(add, "capabilities in add")
        caps.update(normalize_capability_inputs(add))
    removed = set(expand_capability_exclusions(normalize_capability_inputs(remove))) if remove else set()

    changed = True
    while changed:
        changed = False
        for cap in list(caps):
            for dep in CAPABILITY_DEPS.get(cap, []):
                if dep not in caps:
                    caps.add(dep)
                    changed = True

    caps -= removed

    if not caps:
        raise ValueError("Resolved capability set is empty")

    modifiers_present = caps & MODIFIER_ONLY_CAPABILITIES
    if modifiers_present and not (caps & BASE_CAPABILITIES):
        raise ValueError(
            f"Modifier capabilities {sorted(modifiers_present)} require at least one "
            f"base capability ({', '.join(sorted(BASE_CAPABILITIES))})"
        )

    return sorted(caps, key=lambda c: CANONICAL_ORDER.index(c) if c in CANONICAL_ORDER else len(CANONICAL_ORDER))


def expand_capability_exclusions(capabilities: list[str]) -> list[str]:
    excluded = set(capabilities)
    changed = True
    while changed:
        changed = False
        for capability, deps in CAPABILITY_DEPS.items():
            if capability not in excluded and any(dep in excluded for dep in deps):
                excluded.add(capability)
                changed = True
    return sorted(excluded, key=lambda c: CANONICAL_ORDER.index(c) if c in CANONICAL_ORDER else len(CANONICAL_ORDER))


def build_extraction_prompt(
    text: str,
    *,
    capabilities: list[Any] | None = None,
    profile: str | None = None,
    add: list[Any] | None = None,
    remove: list[Any] | None = None,
    categories: list[str] | None = None,
    source_type: str | None = None,
    date: str | None = None,
    stage: str | None = None,
    extracted_at: str | None = None,
    compact: bool | None = None,
) -> str:
    if capabilities is not None and profile is not None:
        raise ValueError("Cannot specify both capabilities and profile")

    resolved = resolve_capabilities(
        capabilities=capabilities,
        profile=profile,
        add=add,
        remove=remove,
    )

    template_ctx: dict[str, Any] = {
        "text": text,
        "categories": categories,
        "source_type": source_type,
        "date": date,
    }

    parts: list[str] = []

    preamble = _render_template(_load_fragment("preamble"), template_ctx)
    parts.append(preamble.strip())

    for cap in resolved:
        fragment = _render_template(_load_fragment(cap), template_ctx)
        parts.append(fragment.rstrip())

    rules_section: list[str] = []
    for cap in resolved:
        rule = CAPABILITY_RULES.get(cap)
        if rule:
            rules_section.append(rule)
    run_constraint_rules = _build_run_constraint_rules(
        resolved,
        stage=stage,
        extracted_at=extracted_at,
        compact=compact,
    )

    postamble_template = _load_fragment("postamble")
    if rules_section or run_constraint_rules:
        extra_blocks: list[str] = []
        if rules_section:
            extra_blocks.append("\n".join(f"- {r}" for r in rules_section))
        if run_constraint_rules:
            extra_blocks.append(
                "Additional run constraints:\n"
                + "\n".join(f"- {r}" for r in run_constraint_rules)
            )
        extra_rules = "\n".join(extra_blocks)
        postamble_rendered = _render_template(postamble_template, template_ctx).rstrip()
        idx = postamble_rendered.find("\nText:")
        if idx >= 0:
            postamble_rendered = postamble_rendered[:idx] + "\n" + extra_rules + postamble_rendered[idx:]
        parts.append(postamble_rendered)
    else:
        parts.append(_render_template(postamble_template, template_ctx).rstrip())

    return "\n".join(parts) + "\n"
